From: kkuo <kat.kuo@canonical.com>
Date: Mon, 5 Jan 2026 16:31:08 -0500
Subject: telemetry: Repurpose diagnostics panel for all telemetry
 applications

This patch renames and repurposes the diagnostics panel to be a catchall
for telemetry-related features.

The Diagnostics panel is renamed to Telemetry while maintaining the same
ABRT and Whoopsie functionality and modularization. To reduce the patch
size, the Diagnostics panel code is deactivated, not removed.

Additional modular integration with Ubuntu Insights via the insights
build argument is also added. This integration allows for user consent
to be managed in a centralized and fine-grained manner. Additionally,
functionality to generate and show an example report is also present.
---
 meson.build                                        |   9 +
 meson_options.txt                                  |   1 +
 panels/common/gnome-control-center.rules.in        |   3 +-
 panels/privacy/cc-privacy-panel.blp                |  12 +-
 panels/privacy/cc-privacy-panel.c                  |   4 +-
 panels/privacy/gnome-privacy-panel.desktop.in      |   2 +-
 panels/privacy/meson.build                         |  30 +-
 ...vacy.gresource.xml => privacy.gresource.xml.in} |   5 +-
 .../telemetry/cc-telemetry-insights-example.blp    |  47 ++
 .../telemetry/cc-telemetry-insights-example.c      | 102 +++
 .../telemetry/cc-telemetry-insights-example.h}     |  13 +-
 panels/privacy/telemetry/cc-telemetry-page.blp     | 115 ++++
 panels/privacy/telemetry/cc-telemetry-page.c       | 765 +++++++++++++++++++++
 .../cc-telemetry-page.h}                           |  10 +-
 .../gnome-control-center-insights-helper.c         |  62 ++
 panels/privacy/telemetry/meson.build               |  42 ++
 ...rolcenter.insights-system-opt-out.policy.in.in} |  11 +-
 panels/privacy/telemetry/telemetry.css             |   8 +
 po/POTFILES.in                                     |   6 +
 19 files changed, 1217 insertions(+), 30 deletions(-)
 rename panels/privacy/{privacy.gresource.xml => privacy.gresource.xml.in} (88%)
 create mode 100644 panels/privacy/telemetry/cc-telemetry-insights-example.blp
 create mode 100644 panels/privacy/telemetry/cc-telemetry-insights-example.c
 copy panels/{mouse/cc-mouse-test.h => privacy/telemetry/cc-telemetry-insights-example.h} (62%)
 create mode 100644 panels/privacy/telemetry/cc-telemetry-page.blp
 create mode 100644 panels/privacy/telemetry/cc-telemetry-page.c
 copy panels/privacy/{diagnostics/cc-diagnostics-page.h => telemetry/cc-telemetry-page.h} (73%)
 create mode 100644 panels/privacy/telemetry/gnome-control-center-insights-helper.c
 create mode 100644 panels/privacy/telemetry/meson.build
 copy panels/{keyboard/org.gnome.controlcenter.keyboard.policy.in.in => privacy/telemetry/org.gnome.controlcenter.insights-system-opt-out.policy.in.in} (60%)
 create mode 100644 panels/privacy/telemetry/telemetry.css

diff --git a/meson.build b/meson.build
index dbbb5ba..423bad6 100644
--- a/meson.build
+++ b/meson.build
@@ -355,6 +355,14 @@ endif
 config_h.set('HAVE_WHOOPSIE', enable_whoopsie,
              description: 'Define to 1 if Whoopsie is enabled')
 
+# Insights (Ubuntu telemetry service)
+enable_insights = get_option('insights')
+if enable_insights
+  insights_dep = dependency('insights', version: '>= 0.8.1')
+endif
+config_h.set('HAVE_INSIGHTS', enable_insights,
+             description: 'Define to 1 if Ubuntu Insights is enabled')
+
 if host_is_linux
   # ModemManager
   mm_dep =  dependency('mm-glib', version: '>= 0.7')
@@ -462,6 +470,7 @@ summary({
     'Snap': enable_snap,
     'Malcontent': enable_malcontent,
     'Whoopsie': enable_whoopsie,
+    'Insights': enable_insights,
   },
   section: 'Optional Dependencies',
   bool_yn: true,
diff --git a/meson_options.txt b/meson_options.txt
index 20de3e1..b222b2a 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -10,3 +10,4 @@ option('ssh_socket_activation', type: 'boolean', value: false, description: 'bui
 option('distributor_logo', type: 'string', description: 'absolute path to distributor logo for the About panel')
 option('dark_mode_distributor_logo', type: 'string', description: 'absolute path to distributor logo dark mode variant')
 option('whoopsie', type: 'boolean', value: false, description: 'build with Whoopsie support')
+option('insights', type: 'boolean', value: false, description: 'build with Ubuntu Insights support')
diff --git a/panels/common/gnome-control-center.rules.in b/panels/common/gnome-control-center.rules.in
index dc0cfad..06ff442 100644
--- a/panels/common/gnome-control-center.rules.in
+++ b/panels/common/gnome-control-center.rules.in
@@ -6,7 +6,8 @@ polkit.addRule(function(action, subject) {
 	     action.id == "org.freedesktop.hostname1.set-static-hostname" ||
 	     action.id == "org.freedesktop.hostname1.set-hostname" ||
 	     action.id == "org.gnome.controlcenter.set-keyboard" ||
-	     action.id == "org.gnome.controlcenter.datetime.configure") &&
+	     action.id == "org.gnome.controlcenter.datetime.configure" ||
+	     action.id == "org.gnome.controlcenter.insights-system-opt-out") &&
 	    subject.local &&
 	    subject.active &&
 	    subject.isInGroup ("@PRIVILEGED_GROUP@")) {
diff --git a/panels/privacy/cc-privacy-panel.blp b/panels/privacy/cc-privacy-panel.blp
index e7daa37..1685bdf 100644
--- a/panels/privacy/cc-privacy-panel.blp
+++ b/panels/privacy/cc-privacy-panel.blp
@@ -41,13 +41,13 @@ template $CcPrivacyPanel: $CcPanel {
         }
 
         $CcListRow {
-          title: _("_Diagnostics");
-          subtitle: _("Automatic problem reporting");
-          icon-name: "org.gnome.Settings-device-diagnostics-symbolic";
+          title: _("Tele_metry");
+          subtitle: _("Control error and system reporting");
+          icon-name: "diagnostics-symbolic";
           show-arrow: true;
           action-name: "navigation.push";
-          action-target: "'diagnostics'";
-          visible: bind diagnostics_page.visible;
+          action-target: "'telemetry'";
+          visible: bind telemetry_page.visible;
         }
 
         $CcListRow {
@@ -105,7 +105,7 @@ template $CcPrivacyPanel: $CcPanel {
   $CcUsagePage {}
 
   [subpage]
-  $CcDiagnosticsPage diagnostics_page {}
+  $CcTelemetryPage telemetry_page {}
 
   [subpage]
   $CcConnectivityPage {}
diff --git a/panels/privacy/cc-privacy-panel.c b/panels/privacy/cc-privacy-panel.c
index 3a295dc..f348c76 100644
--- a/panels/privacy/cc-privacy-panel.c
+++ b/panels/privacy/cc-privacy-panel.c
@@ -29,7 +29,7 @@
 #include "cc-list-row.h"
 #include "cc-privacy-resources.h"
 #include "connectivity/cc-connectivity-page.h"
-#include "diagnostics/cc-diagnostics-page.h"
+#include "telemetry/cc-telemetry-page.h"
 #include "firmware-security/cc-firmware-security-page.h"
 #include "location/cc-location-page.h"
 #include "screen/cc-screen-page.h"
@@ -81,7 +81,7 @@ cc_privacy_panel_class_init (CcPrivacyPanelClass *klass)
 
     g_type_ensure (CC_TYPE_CAMERA_PAGE);
     g_type_ensure (CC_TYPE_CONNECTIVITY_PAGE);
-    g_type_ensure (CC_TYPE_DIAGNOSTICS_PAGE);
+    g_type_ensure (CC_TYPE_TELEMETRY_PAGE);
     g_type_ensure (CC_TYPE_FIRMWARE_SECURITY_PAGE);
     g_type_ensure (CC_TYPE_LOCATION_PAGE);
     g_type_ensure (CC_TYPE_SCREEN_PAGE);
diff --git a/panels/privacy/gnome-privacy-panel.desktop.in b/panels/privacy/gnome-privacy-panel.desktop.in
index 3704b2b..e44caf3 100644
--- a/panels/privacy/gnome-privacy-panel.desktop.in
+++ b/panels/privacy/gnome-privacy-panel.desktop.in
@@ -11,4 +11,4 @@ StartupNotify=true
 Categories=GNOME;GTK;Settings;DesktopSettings;X-GNOME-Settings-Panel;X-GNOME-AccountSettings;
 OnlyShowIn=GNOME;
 # Translators: Search terms to find the Privacy panel. Do NOT translate or localize the semicolons! The list MUST also end with a semicolon!
-Keywords=Privacy;Screen;Lock;Private;Usage;Recent;History;Files;Temporary;Tmp;Trash;Purge;Retain;Diagnostics;Crash;Location;Gps;Camera;Photos;Video;Webcam;Recording;Security;Firmware;Thunderbolt;Network;error;bug;whoopsie;launchpad;issue;
+Keywords=Privacy;Screen;Lock;Private;Usage;Recent;History;Files;Telemetry;Temporary;Tmp;Trash;Purge;Retain;Diagnostics;Crash;Location;Gps;Camera;Photos;Video;Webcam;Recording;Security;Firmware;Thunderbolt;Network;error;bug;whoopsie;launchpad;issue;
diff --git a/panels/privacy/meson.build b/panels/privacy/meson.build
index 03078c7..b71dcce 100644
--- a/panels/privacy/meson.build
+++ b/panels/privacy/meson.build
@@ -15,7 +15,7 @@ blueprints = []
 subdir('bolt')
 subdir('camera')
 subdir('connectivity')
-subdir('diagnostics')
+subdir('telemetry')
 subdir('firmware-security')
 subdir('icons')
 subdir('location')
@@ -40,7 +40,7 @@ blueprints += custom_target(
 sources = files(
   'camera/cc-camera-page.c',
   'connectivity/cc-connectivity-page.c',
-  'diagnostics/cc-diagnostics-page.c',
+  'telemetry/cc-telemetry-page.c',
   'firmware-security/cc-firmware-security-boot-dialog.c',
   'firmware-security/cc-firmware-security-dialog.c',
   'firmware-security/cc-firmware-security-help-dialog.c',
@@ -52,6 +52,10 @@ sources = files(
   'usage/cc-usage-page.c',
 )
 
+if enable_insights
+sources += files(  'telemetry/cc-telemetry-insights-example.c')
+endif
+
 if host_is_linux_not_s390
   sources += files(
     'bolt/bolt-client.c',
@@ -78,16 +82,30 @@ if host_is_linux_not_s390
   )
 endif
 
+resource_conf = configuration_data()
+telemetry_ui = ''
+if enable_insights
+  telemetry_ui = '    <file preprocess="xml-stripblanks">telemetry/cc-telemetry-insights-example.ui</file>'
+endif
+resource_conf.set('telemetry_insights_example_ui', telemetry_ui)
+
+resource_file = configure_file(
+  input: cappletname + '.gresource.xml.in',
+  output: cappletname + '.gresource.xml',
+  configuration: resource_conf
+)
+
 sources += gnome.compile_resources(
   'cc-' + cappletname + '-resources',
-  cappletname + '.gresource.xml',
+  resource_file,
   c_name: 'cc_' + cappletname,
   export: true,
   dependencies: blueprints,
 )
 
 cflags += [
-  '-DDATADIR="@0@"'.format(control_center_datadir)
+  '-DDATADIR="@0@"'.format(control_center_datadir),
+  '-DLIBEXECDIR="@0@"'.format(control_center_libexecdir),
 ]
 
 deps = common_deps + [
@@ -107,6 +125,10 @@ if enable_whoopsie
   deps += whoopsie_dep
 endif
 
+if enable_insights
+  deps += insights_dep
+endif
+
 panels_libs += static_library(
   cappletname,
   sources: sources,
diff --git a/panels/privacy/privacy.gresource.xml b/panels/privacy/privacy.gresource.xml.in
similarity index 88%
rename from panels/privacy/privacy.gresource.xml
rename to panels/privacy/privacy.gresource.xml.in
index 0670aa6..de0da0d 100644
--- a/panels/privacy/privacy.gresource.xml
+++ b/panels/privacy/privacy.gresource.xml.in
@@ -6,7 +6,8 @@
     <file preprocess="xml-stripblanks">bolt/cc-bolt-page.ui</file>
     <file preprocess="xml-stripblanks">camera/cc-camera-page.ui</file>
     <file preprocess="xml-stripblanks">connectivity/cc-connectivity-page.ui</file>
-    <file preprocess="xml-stripblanks">diagnostics/cc-diagnostics-page.ui</file>
+    <file preprocess="xml-stripblanks">telemetry/cc-telemetry-page.ui</file>
+@telemetry_insights_example_ui@
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-boot-dialog.ui</file>
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-dialog.ui</file>
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-help-dialog.ui</file>
@@ -15,5 +16,7 @@
     <file preprocess="xml-stripblanks">cc-privacy-panel.ui</file>
     <file preprocess="xml-stripblanks">screen/cc-screen-page.ui</file>
     <file preprocess="xml-stripblanks">usage/cc-usage-page.ui</file>
+
+    <file>telemetry/telemetry.css</file>
   </gresource>
 </gresources>
diff --git a/panels/privacy/telemetry/cc-telemetry-insights-example.blp b/panels/privacy/telemetry/cc-telemetry-insights-example.blp
new file mode 100644
index 0000000..3f5ce2f
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.blp
@@ -0,0 +1,47 @@
+using Gtk 4.0;
+using Adw 1;
+
+template $CcTelemetryInsightsExample: Adw.Window {
+  title: _("System Report");
+  modal: true;
+  width-request: 535;
+  height-request: 600;
+
+  content: Adw.ToolbarView {
+    [top]
+    Adw.HeaderBar {}
+
+    content: ScrolledWindow {
+      hscrollbar-policy: never;
+
+      child: Box {
+        orientation: vertical;
+        spacing: 12;
+        margin-top: 29;
+        margin-bottom: 29;
+        margin-start: 24;
+        margin-end: 24;
+
+        Label subtitle_label {
+          wrap: true;
+          wrap-mode: word_char;
+          valign: center;
+          halign: start;
+        }
+
+        TextView report_view {
+          editable: false;
+          cursor-visible: false;
+          monospace: true;
+          wrap-mode: word_char;
+          left-margin: 12;
+          right-margin: 12;
+          top-margin: 12;
+          bottom-margin: 12;
+
+          buffer: Gtk.TextBuffer {};
+        }
+      };
+    };
+  };
+}
diff --git a/panels/privacy/telemetry/cc-telemetry-insights-example.c b/panels/privacy/telemetry/cc-telemetry-insights-example.c
new file mode 100644
index 0000000..c3c36b0
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.c
@@ -0,0 +1,102 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2026 Canonical Ltd.
+ *
+ * Written by: Kat Kuo <kat.kuo@canonical.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+
+#include <glib/gi18n.h>
+
+#include "cc-telemetry-insights-example.h"
+#include "glib.h"
+#include "gtk/gtk.h"
+#include "gtk/gtkshortcut.h"
+
+#define DOCUMENTATION_URL "https://github.com/ubuntu/ubuntu-insights#example-insights-reports"
+#define INSIGHTS_CACHE_SUBDIR ".cache/ubuntu-insights"
+
+struct _CcTelemetryInsightsExample
+{
+  AdwWindow parent_instance;
+
+  GtkWidget   *subtitle_label;
+  GtkTextView *report_view;
+};
+
+G_DEFINE_TYPE (CcTelemetryInsightsExample, cc_telemetry_insights_example, ADW_TYPE_WINDOW)
+
+static void
+cc_telemetry_insights_example_class_init (CcTelemetryInsightsExampleClass *klass)
+{
+  GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+  gtk_widget_class_add_binding_action (widget_class, GDK_KEY_Escape, 0, "window.close", NULL);
+
+  gtk_widget_class_set_template_from_resource (widget_class,
+                                               "/org/gnome/control-center/privacy/telemetry/cc-telemetry-insights-example.ui");
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryInsightsExample, subtitle_label);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryInsightsExample, report_view);
+}
+
+static void
+cc_telemetry_insights_example_init (CcTelemetryInsightsExample *self)
+{
+  g_autofree gchar *path = NULL;
+  g_autofree gchar *url = NULL;
+  g_autofree gchar *subtitle = NULL;
+  g_autofree gchar *cache_path = NULL;
+  g_autofree gchar *cache_uri = NULL;
+  g_autofree gchar *display_path = NULL;
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+
+  cache_path = g_build_filename (g_get_home_dir (), INSIGHTS_CACHE_SUBDIR, NULL);
+  display_path = g_strdup_printf ("~/%s", INSIGHTS_CACHE_SUBDIR);
+
+  if (g_file_test (cache_path, G_FILE_TEST_IS_DIR))
+    {
+      cache_uri = g_filename_to_uri (cache_path, NULL, NULL);
+      path = g_strdup_printf ("<a href=\"%s\"><span face=\"monospace\" allow_breaks=\"false\">%s</span></a>",
+                              cache_uri,
+                              display_path);
+    }
+  else
+    {
+      path = g_strdup_printf ("<span face=\"monospace\" allow_breaks=\"false\">%s</span>",
+                              display_path);
+    }
+
+  /* translators: The '%s' is a file path to the local Ubuntu Insights cache directory */
+  subtitle = g_strdup_printf (_ ("This is a system report example with real data from this computer. "
+                                 "System reports are temporarily stored in %s before being sent to Canonical."),
+                              path);
+  gtk_label_set_markup (GTK_LABEL (self->subtitle_label), subtitle);
+}
+
+GtkWidget *
+cc_telemetry_insights_example_new (const gchar *report)
+{
+  CcTelemetryInsightsExample *self;
+  self = g_object_new (CC_TYPE_TELEMETRY_INSIGHTS_EXAMPLE, NULL);
+
+  gtk_text_buffer_set_text (gtk_text_view_get_buffer (self->report_view),
+                            report,
+                            -1);
+  return GTK_WIDGET (self);
+}
diff --git a/panels/mouse/cc-mouse-test.h b/panels/privacy/telemetry/cc-telemetry-insights-example.h
similarity index 62%
copy from panels/mouse/cc-mouse-test.h
copy to panels/privacy/telemetry/cc-telemetry-insights-example.h
index a0e2126..bb5f6cc 100644
--- a/panels/mouse/cc-mouse-test.h
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.h
@@ -1,8 +1,8 @@
-/* -*- mode: c; style: linux -*-
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
  *
- * Copyright (C) 2012 Red Hat, Inc.
+ * Copyright (C) 2026 Canonical Ltd.
  *
- * Written by: Ondrej Holy <oholy@redhat.com>
+ * Written by: Kat Kuo <kat.kuo@canonical.com>
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -24,8 +24,9 @@
 
 G_BEGIN_DECLS
 
-#define CC_TYPE_MOUSE_TEST (cc_mouse_test_get_type ())
-G_DECLARE_FINAL_TYPE (CcMouseTest, cc_mouse_test, CC, MOUSE_TEST, AdwWindow);
-GtkWidget *cc_mouse_test_new (void);
+#define CC_TYPE_TELEMETRY_INSIGHTS_EXAMPLE (cc_telemetry_insights_example_get_type ())
+G_DECLARE_FINAL_TYPE (CcTelemetryInsightsExample, cc_telemetry_insights_example, CC, TELEMETRY_INSIGHTS_EXAMPLE, AdwWindow)
+
+GtkWidget *cc_telemetry_insights_example_new (const gchar *report);
 
 G_END_DECLS
diff --git a/panels/privacy/telemetry/cc-telemetry-page.blp b/panels/privacy/telemetry/cc-telemetry-page.blp
new file mode 100644
index 0000000..6fcf660
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-page.blp
@@ -0,0 +1,115 @@
+using Gtk 4.0;
+using Adw 1;
+
+template $CcTelemetryPage: Adw.NavigationPage {
+  title: _("Telemetry");
+  tag: "telemetry";
+  visible: false;
+
+  child: Adw.ToolbarView {
+    [top]
+    Adw.HeaderBar {}
+
+    content: Box {
+      orientation: vertical;
+
+      $CcPermissionInfobar permission_infobar {
+        visible: false;
+      }
+
+      Adw.PreferencesPage {
+        Adw.PreferencesGroup insights_group {
+          title: _("System Reporting");
+          visible: false;
+
+          Adw.SwitchRow insights_linux_row {
+            title: _("Send a system report every month");
+            sensitive: bind insights_system_opt_out_row.active inverted;
+            has-tooltip: bind insights_system_opt_out_row.active;
+            tooltip-text: _("System reporting is turned off for all users");
+            notify::active => $insights_switch_row_changed_cb();
+          }
+
+          Adw.SwitchRow insights_upgrade_row {
+            title: _("Send a system report when upgrading to a new Ubuntu version");
+            sensitive: bind insights_system_opt_out_row.active inverted;
+            has-tooltip: bind insights_system_opt_out_row.active;
+            tooltip-text: _("System reporting is turned off for all users");
+            notify::active => $insights_switch_row_changed_cb();
+          }
+        }
+
+        Adw.PreferencesGroup insights_system_group {
+          visible: bind insights_group.visible;
+
+          Adw.SwitchRow insights_system_opt_out_row {
+            title: _("Opt out of system reporting for all users");
+            has-tooltip: bind insights_system_opt_out_row.sensitive inverted;
+            tooltip-text: _("Unlock to Change This Setting");
+            notify::active => $insights_system_opt_out_row_changed_cb(template);
+          }
+        }
+
+        Adw.PreferencesGroup insights_buttons_group {
+          separate-rows: true;
+          visible: bind insights_group.visible;
+
+          Adw.ButtonRow insights_example_report_button_row {
+            title: _("_Show Report");
+            use-underline: true;
+            end-icon-name: "go-next-symbolic";
+            activated => $insights_example_report_button_row_activated_cb(template);
+          }
+
+          Adw.ButtonRow insights_cancel_report_button_row {
+            title: _("Cancel Report Generation");
+            visible: false;
+            activated => $insights_cancel_report_button_row_activated_cb(template);
+          }
+
+          LinkButton insights_learn_more_link_button {
+            label: _("Learn more about system reporting");
+            uri: "https://canonical.com/legal/systems-information-notice";
+            margin-top: 8;
+            halign: start;
+
+            styles [
+              "dim",
+              "body",
+              "link-label",
+            ]
+          }
+        }
+
+        Adw.PreferencesGroup diagnostics_group {
+          title: _("Error Reporting");
+
+          Adw.SwitchRow abrt_row {
+            title: _("_Automatic Error Reporting");
+            use-underline: true;
+          }
+
+          Adw.ComboRow whoopsie_combo_row {
+            title: _("Send error reports to Canonical");
+            visible: false;
+            use-underline: true;
+            notify::selected-item => $whoopsie_combo_row_changed_cb(template);
+          }
+
+          LinkButton diagnostics_learn_more_link_button {
+            label: _("Learn more about error reporting");
+            visible: bind $string_is_not_empty(diagnostics_learn_more_link_button.uri) as <bool>;
+            margin-top: 8;
+            halign: start;
+
+            styles [
+              "dim",
+              "body",
+              "link-label",
+            ]
+          }
+        }
+      }
+    };
+  };
+}
diff --git a/panels/privacy/telemetry/cc-telemetry-page.c b/panels/privacy/telemetry/cc-telemetry-page.c
new file mode 100644
index 0000000..70fabf9
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-page.c
@@ -0,0 +1,765 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2026 Canonical Ltd.
+ * Copyright (C) 2018 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author(s):
+ *   Kat Kuo <kat.kuo@canonical.com>
+ *   Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include <config.h>
+
+#include "cc-telemetry-page.h"
+#include "cc-util.h"
+#include "gtk/gtk.h"
+#include "shell/cc-application.h"
+
+#include <gio/gdesktopappinfo.h>
+#include <glib/gi18n.h>
+#include <stdio.h>
+
+#include "cc-permission-infobar.h"
+
+#ifdef HAVE_INSIGHTS
+#include <polkit/polkit.h>
+#endif
+
+#ifdef HAVE_WHOOPSIE
+#include <whoopsie-preferences/libwhoopsie-preferences.h>
+#endif
+
+#ifdef HAVE_INSIGHTS
+#include "cc-telemetry-insights-example.h"
+#include <insights/insights.h>
+#include <insights/types.h>
+#endif
+
+struct _CcTelemetryPage
+{
+  AdwNavigationPage parent_instance;
+
+  AdwPreferencesGroup *diagnostics_group;
+  AdwSwitchRow        *abrt_row;
+
+#ifdef HAVE_WHOOPSIE
+  AdwComboRow         *whoopsie_combo_row;
+  WhoopsiePreferences *whoopsie;
+#endif
+  GtkLinkButton *diagnostics_learn_more_link_button;
+
+  AdwPreferencesGroup *insights_group;
+  AdwPreferencesGroup *insights_buttons_group;
+
+#ifdef HAVE_INSIGHTS
+  AdwSwitchRow *insights_linux_row;
+  AdwSwitchRow *insights_upgrade_row;
+  AdwSwitchRow *insights_system_opt_out_row;
+
+  AdwButtonRow *insights_example_report_button_row;
+  AdwButtonRow *insights_cancel_report_button_row;
+
+  CcPermissionInfobar *permission_infobar;
+  GPermission         *permission;
+#endif
+
+  GCancellable *cancellable;
+
+#ifdef HAVE_INSIGHTS
+  GCancellable *insights_cancellable;
+#endif
+
+  GSettings *privacy_settings;
+};
+
+G_DEFINE_TYPE (CcTelemetryPage, cc_telemetry_page, ADW_TYPE_NAVIGATION_PAGE)
+
+#ifdef HAVE_INSIGHTS
+#define INSIGHTS_SYSTEM_OPT_OUT_PERMISSION "org.gnome.controlcenter.insights-system-opt-out"
+#endif
+
+/* Static init function */
+
+static gboolean
+string_is_not_empty (CcTelemetryPage *self,
+                     const char      *string)
+{
+  return string && string[0];
+}
+
+static void
+abrt_appeared_cb (GDBusConnection *connection,
+                  const gchar     *name,
+                  const gchar     *name_owner,
+                  gpointer         user_data)
+{
+  g_autofree gchar *whoopsie_url = NULL;
+  CcTelemetryPage  *self = CC_TELEMETRY_PAGE (user_data);
+
+  g_debug ("ABRT appeared");
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->diagnostics_group), TRUE);
+}
+
+static void
+abrt_vanished_cb (GDBusConnection *connection,
+                  const gchar     *name,
+                  gpointer         user_data)
+{
+  g_debug ("ABRT vanished");
+#ifndef HAVE_WHOOPSIE
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  gtk_widget_set_visible (GTK_WIDGET (self->diagnostics_group), FALSE);
+
+#ifndef HAVE_INSIGHTS
+  gtk_widget_set_visible (GTK_WIDGET (self), FALSE);
+#endif
+#endif
+}
+
+#ifdef HAVE_WHOOPSIE
+typedef enum
+{
+  WHOOPSIE_BUTTON_SETTING_NEVER,
+  WHOOPSIE_BUTTON_SETTING_AUTO,
+  WHOOPSIE_BUTTON_SETTING_MANUAL,
+} WhoopsieButtonSettingType;
+
+static void
+whoopsie_set_report_crashes_done (GObject      *source_object,
+                                  GAsyncResult *res,
+                                  gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autoptr (GError) error = NULL;
+
+  if (!whoopsie_preferences_call_set_report_crashes_finish (self->whoopsie, res, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to toggle crash reporting: %s", error->message);
+    }
+}
+
+static void
+whoopsie_set_report_crashes_mode_done (GObject      *source_object,
+                                       GAsyncResult *res,
+                                       gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autoptr (GError) error = NULL;
+
+  if (!whoopsie_preferences_call_set_automatically_report_crashes_finish (self->whoopsie, res, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to toggle crash reporting mode: %s", error->message);
+    }
+}
+
+static void
+whoopsie_combo_row_changed_cb (CcTelemetryPage *self)
+{
+  g_autoptr (GObject) item = NULL;
+  GListModel *model;
+  gint        selected_index;
+  gint        value;
+
+  model = adw_combo_row_get_model (self->whoopsie_combo_row);
+  selected_index = adw_combo_row_get_selected (self->whoopsie_combo_row);
+  if (selected_index == -1)
+    return;
+
+  item = g_list_model_get_item (model, selected_index);
+  value = GPOINTER_TO_UINT (g_object_get_data (G_OBJECT (item), "value"));
+
+  whoopsie_preferences_call_set_report_crashes (self->whoopsie,
+                                                value != WHOOPSIE_BUTTON_SETTING_NEVER,
+                                                self->cancellable,
+                                                whoopsie_set_report_crashes_done,
+                                                self);
+
+  whoopsie_preferences_call_set_automatically_report_crashes (self->whoopsie,
+                                                              value == WHOOPSIE_BUTTON_SETTING_AUTO,
+                                                              self->cancellable,
+                                                              whoopsie_set_report_crashes_mode_done,
+                                                              self);
+}
+
+static void
+set_value_for_combo_row (AdwComboRow *combo_row, gint value)
+{
+  g_autoptr (GObject) new_item = NULL;
+  GListModel *model;
+  guint       i;
+
+  /* try to make the UI match the setting */
+  model = adw_combo_row_get_model (combo_row);
+  for (i = 0; i < g_list_model_get_n_items (model); i++)
+    {
+      g_autoptr (GObject) item = g_list_model_get_item (model, i);
+      gint value_tmp = GPOINTER_TO_UINT (g_object_get_data (item, "value"));
+      if (value_tmp == value)
+        {
+          adw_combo_row_set_selected (combo_row, i);
+          return;
+        }
+    }
+}
+
+static void
+whoopsie_properties_changed (CcTelemetryPage *self)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (G_DBUS_PROXY (self->whoopsie));
+
+  /* Bail out if dbus service has gone away */
+  if (G_UNLIKELY (!name_owner))
+    return;
+
+  WhoopsieButtonSettingType value = WHOOPSIE_BUTTON_SETTING_NEVER;
+
+  if (whoopsie_preferences_get_automatically_report_crashes (self->whoopsie))
+    value = WHOOPSIE_BUTTON_SETTING_AUTO;
+  else if (whoopsie_preferences_get_report_crashes (self->whoopsie))
+    value = WHOOPSIE_BUTTON_SETTING_MANUAL;
+
+  g_signal_handlers_block_by_func (self->whoopsie_combo_row, whoopsie_combo_row_changed_cb, self);
+  set_value_for_combo_row (self->whoopsie_combo_row, value);
+  g_signal_handlers_unblock_by_func (self->whoopsie_combo_row, whoopsie_combo_row_changed_cb, self);
+}
+
+static void
+whoopsie_bus_owner_changed (GDBusProxy *proxy)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (proxy);
+
+  if (!name_owner)
+    {
+      /* Trigger the service to start again, as it has just been stopped by a timeout
+       * don't bother on waiting for the call results.
+       */
+      whoopsie_preferences_call_get_identifier (WHOOPSIE_PREFERENCES (proxy), NULL, NULL, NULL);
+    }
+}
+
+static void
+populate_whoopsie_button_row (AdwComboRow *combo_row)
+{
+  g_autoptr (GtkStringList) string_list = NULL;
+  struct
+  {
+    char                     *name;
+    WhoopsieButtonSettingType value;
+  } actions[] = {
+    { NC_ ("Whoopsie error reporting", "Never"), WHOOPSIE_BUTTON_SETTING_NEVER },
+    { NC_ ("Whoopsie error reporting", "Automatic"), WHOOPSIE_BUTTON_SETTING_AUTO },
+    { NC_ ("Whoopsie error reporting", "Manual"), WHOOPSIE_BUTTON_SETTING_MANUAL },
+  };
+  guint item_index = 0;
+  guint i;
+
+  string_list = gtk_string_list_new (NULL);
+  for (i = 0; i < G_N_ELEMENTS (actions); i++)
+    {
+      g_autoptr (GObject) item = NULL;
+
+      gtk_string_list_append (string_list, _ (actions[i].name));
+
+      item = g_list_model_get_item (G_LIST_MODEL (string_list), item_index++);
+      g_object_set_data (item, "value", GUINT_TO_POINTER (actions[i].value));
+    }
+
+  adw_combo_row_set_model (combo_row, G_LIST_MODEL (string_list));
+}
+
+static void
+on_new_whoopsie_proxy_cb (GObject      *object,
+                          GAsyncResult *res,
+                          gpointer      data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (data);
+  g_autoptr (GError) error = NULL;
+
+  self->whoopsie = whoopsie_preferences_proxy_new_for_bus_finish (res, &error);
+
+  if (error)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_warning ("Failed to get whoopsie proxy: %s", error->message);
+
+      gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), FALSE);
+      gtk_widget_set_visible (GTK_WIDGET (self->abrt_row), TRUE);
+      return;
+    }
+
+  g_debug ("Whoopsie available");
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+  gtk_widget_set_sensitive (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+
+  g_signal_handlers_block_by_func (self->whoopsie_combo_row,
+                                   whoopsie_combo_row_changed_cb, self);
+  populate_whoopsie_button_row (self->whoopsie_combo_row);
+  g_signal_handlers_unblock_by_func (self->whoopsie_combo_row,
+                                     whoopsie_combo_row_changed_cb, self);
+
+  g_signal_connect_object (self->whoopsie, "g-properties-changed",
+                           G_CALLBACK (whoopsie_properties_changed),
+                           self, G_CONNECT_SWAPPED);
+
+  g_signal_connect_object (self->whoopsie, "notify::g-name-owner",
+                           G_CALLBACK (whoopsie_bus_owner_changed),
+                           self, 0);
+
+  whoopsie_properties_changed (self);
+}
+#else
+static void
+whoopsie_combo_row_changed_cb (CcTelemetryPage *self)
+{
+}
+#endif
+
+#ifdef HAVE_INSIGHTS
+static void update_insights_row_state (CcTelemetryPage *self,
+                                       AdwSwitchRow    *row,
+                                       gboolean         system_opt_out);
+static void insights_system_opt_out_row_changed_cb (CcTelemetryPage *self);
+static void sync_insights_switches (CcTelemetryPage *self);
+
+static void
+insights_switch_row_changed_cb (GObject    *object,
+                                GParamSpec *pspec,
+                                gpointer    user_data)
+{
+  AdwSwitchRow    *row = ADW_SWITCH_ROW (object);
+  gboolean         active;
+  const gchar     *source;
+  g_autofree char *error_msg = NULL;
+
+  active = adw_switch_row_get_active (row);
+  source = g_object_get_data (G_OBJECT (row), "insights-source");
+
+  if (!source)
+    return; /* Shouldn't happen */
+
+  error_msg = insights_set_consent_state (NULL,
+                                          source,
+                                          active ? INSIGHTS_CONSENT_TRUE : INSIGHTS_CONSENT_FALSE);
+  if (error_msg != NULL)
+    {
+      g_warning ("Failed to set insights consent state for source '%s': %s", source, error_msg);
+      update_insights_row_state (CC_TELEMETRY_PAGE (user_data),
+                                 row,
+                                 insights_get_system_opt_out_state (NULL));
+    }
+}
+
+static void
+update_insights_row_state (CcTelemetryPage *self,
+                           AdwSwitchRow    *row,
+                           gboolean         system_opt_out)
+{
+  const char            *source;
+  insights_consent_state state;
+
+  source = g_object_get_data (G_OBJECT (row), "insights-source");
+  if (!source)
+    return; /* Shouldn't happen */
+
+  state = insights_get_consent_state (NULL, source);
+  g_signal_handlers_block_by_func (row, insights_switch_row_changed_cb, self);
+  adw_switch_row_set_active (row, !system_opt_out && state == INSIGHTS_CONSENT_TRUE);
+  g_signal_handlers_unblock_by_func (row, insights_switch_row_changed_cb, self);
+}
+
+static void
+update_insights_system_opt_out_row (CcTelemetryPage *self)
+{
+  gboolean state;
+
+  state = insights_get_system_opt_out_state (NULL);
+  g_signal_handlers_block_by_func (self->insights_system_opt_out_row,
+                                   insights_system_opt_out_row_changed_cb, self);
+  adw_switch_row_set_active (self->insights_system_opt_out_row, state);
+  g_signal_handlers_unblock_by_func (self->insights_system_opt_out_row,
+                                     insights_system_opt_out_row_changed_cb, self);
+}
+
+static void
+on_set_system_opt_out_cb (GObject      *source_object,
+                          GAsyncResult *res,
+                          gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autoptr (GError) error = NULL;
+
+  if (!g_subprocess_wait_check_finish (G_SUBPROCESS (source_object), res, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to set system opt-out state: %s", error->message);
+    }
+
+  sync_insights_switches (self);
+}
+
+static void
+insights_system_opt_out_row_changed_cb (CcTelemetryPage *self)
+{
+  gboolean                  active;
+  g_autoptr (GError)        error = NULL;
+  g_autoptr (GSubprocess)   subprocess = NULL;
+  g_autoptr (GSubprocessLauncher) launcher = NULL;
+  g_auto (GStrv)            envp = NULL;
+
+  active = adw_switch_row_get_active (self->insights_system_opt_out_row);
+
+  /* The write goes through pkexec so polkit can authorize the
+   * org.gnome.controlcenter.insights-system-opt-out action. */
+  envp = g_get_environ ();
+  envp = g_environ_unsetenv (envp, "SHELL");
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_set_environ (launcher, envp);
+
+  subprocess = g_subprocess_launcher_spawn (launcher,
+                                            &error,
+                                            "/usr/bin/pkexec",
+                                            LIBEXECDIR "/gnome-control-center-insights-helper",
+                                            active ? "true" : "false",
+                                            NULL);
+
+  if (subprocess == NULL)
+    {
+      g_warning ("Failed to launch insights helper: %s", error->message);
+      update_insights_system_opt_out_row (self);
+      sync_insights_switches (self);
+      return;
+    }
+
+  g_subprocess_wait_check_async (subprocess,
+                                 self->cancellable,
+                                 on_set_system_opt_out_cb,
+                                 self);
+}
+
+static void
+sync_insights_switches (CcTelemetryPage *self)
+{
+  gboolean system_opt_out;
+  gboolean allowed;
+
+  update_insights_system_opt_out_row (self);
+
+  allowed = (self->permission == NULL || g_permission_get_allowed (self->permission));
+  gtk_widget_set_sensitive (GTK_WIDGET (self->insights_system_opt_out_row), allowed);
+
+  system_opt_out = insights_get_system_opt_out_state (NULL);
+  update_insights_row_state (self, self->insights_linux_row, system_opt_out);
+  update_insights_row_state (self, self->insights_upgrade_row, system_opt_out);
+}
+
+static void
+on_generate_report_thread (GTask        *task,
+                           gpointer      source_object,
+                           gpointer      task_data,
+                           GCancellable *cancellable)
+{
+  g_autoptr (GError) error = NULL;
+  g_autofree char       *error_msg = NULL;
+  g_autofree char       *report = NULL;
+  insights_compile_flags flags = { 0 };
+
+  error_msg = insights_compile (NULL, &flags, &report);
+  if (error_msg != NULL)
+    {
+      g_task_return_new_error (task, G_IO_ERROR, G_IO_ERROR_FAILED, "%s", error_msg);
+      return;
+    }
+
+  g_task_return_pointer (task, g_steal_pointer (&report), g_free);
+}
+
+static void
+generate_report_cb (GObject      *source_object,
+                    GAsyncResult *res,
+                    gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autofree char *report = NULL;
+  g_autoptr (GError) error = NULL;
+  GtkWidget *window;
+
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), FALSE);
+
+  report = g_task_propagate_pointer (G_TASK (res), &error);
+
+  if (error)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_warning ("Failed to generate example report: %s", error->message);
+      return;
+    }
+
+  window = cc_telemetry_insights_example_new (report);
+  gtk_window_set_transient_for (GTK_WINDOW (window),
+                                GTK_WINDOW (gtk_widget_get_native (GTK_WIDGET (self))));
+  gtk_window_present (GTK_WINDOW (window));
+}
+
+static void
+insights_example_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+  g_autoptr (GTask) task;
+
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), FALSE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), TRUE);
+
+  task = g_task_new (self, self->insights_cancellable, generate_report_cb, self);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_run_in_thread (task, on_generate_report_thread);
+}
+
+static void
+insights_cancel_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+  g_debug ("User cancelled insights report generation");
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), FALSE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), TRUE);
+
+  g_cancellable_cancel (self->insights_cancellable);
+  g_clear_object (&self->insights_cancellable);
+  self->insights_cancellable = g_cancellable_new ();
+}
+
+#else
+static void
+insights_switch_row_changed_cb (GObject    *object,
+                                GParamSpec *pspec,
+                                gpointer    user_data)
+{
+}
+
+static void
+insights_system_opt_out_row_changed_cb (CcTelemetryPage *self)
+{
+}
+
+static void
+insights_example_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+}
+
+static void
+insights_cancel_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+}
+
+#endif
+
+static void
+cc_telemetry_page_finalize (GObject *object)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (object);
+
+  g_clear_object (&self->privacy_settings);
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->cancellable);
+
+#ifdef HAVE_WHOOPSIE
+  g_clear_object (&self->whoopsie);
+#endif
+
+#ifdef HAVE_INSIGHTS
+  g_cancellable_cancel (self->insights_cancellable);
+  g_clear_object (&self->insights_cancellable);
+  g_clear_object (&self->permission);
+#endif
+
+  G_OBJECT_CLASS (cc_telemetry_page_parent_class)->finalize (object);
+}
+
+static void
+cc_telemetry_page_class_init (CcTelemetryPageClass *klass)
+{
+  GObjectClass   *oclass = G_OBJECT_CLASS (klass);
+  GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+  oclass->finalize = cc_telemetry_page_finalize;
+
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/privacy/telemetry/cc-telemetry-page.ui");
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, diagnostics_group);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, abrt_row);
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_group);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_buttons_group);
+
+#ifdef HAVE_WHOOPSIE
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, whoopsie_combo_row);
+#endif
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, diagnostics_learn_more_link_button);
+  gtk_widget_class_bind_template_callback (widget_class, whoopsie_combo_row_changed_cb);
+
+  /* The template references CcPermissionInfobar unconditionally */
+  g_type_ensure (CC_TYPE_PERMISSION_INFOBAR);
+
+#ifdef HAVE_INSIGHTS
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_linux_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_upgrade_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_system_opt_out_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_example_report_button_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_cancel_report_button_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, permission_infobar);
+#endif
+  gtk_widget_class_bind_template_callback (widget_class, insights_switch_row_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, insights_system_opt_out_row_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, insights_example_report_button_row_activated_cb);
+  gtk_widget_class_bind_template_callback (widget_class, insights_cancel_report_button_row_activated_cb);
+
+  gtk_widget_class_bind_template_callback (widget_class, string_is_not_empty);
+}
+
+#ifdef HAVE_INSIGHTS
+static void
+on_unmap_cb (GtkWidget *widget,
+             gpointer   user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (widget);
+
+  if (self->insights_cancellable)
+    {
+      g_cancellable_cancel (self->insights_cancellable);
+      g_object_unref (self->insights_cancellable);
+      self->insights_cancellable = g_cancellable_new ();
+    }
+}
+#endif
+
+static void
+cc_telemetry_page_init (CcTelemetryPage *self)
+{
+  g_autofree gchar *os_name = NULL;
+  g_autofree gchar *whoopsie_url = NULL;
+  g_autofree gchar *whoopsie_msg = NULL;
+  g_autofree gchar *insights_msg = NULL;
+  g_autoptr (GtkCssProvider) provider = NULL;
+  g_autoptr (GError) error = NULL;
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+
+#ifdef HAVE_INSIGHTS
+  g_object_set_data (G_OBJECT (self->insights_linux_row), "insights-source", "linux");
+  g_object_set_data (G_OBJECT (self->insights_upgrade_row), "insights-source", "ubuntu_release_upgrader");
+#endif
+
+  g_bus_watch_name (G_BUS_TYPE_SYSTEM,
+                    "org.freedesktop.problems.daemon",
+                    G_BUS_NAME_WATCHER_FLAGS_NONE,
+                    abrt_appeared_cb,
+                    abrt_vanished_cb,
+                    self,
+                    NULL);
+
+  self->privacy_settings = g_settings_new ("org.gnome.desktop.privacy");
+
+  g_settings_bind (self->privacy_settings, "report-technical-problems",
+                   self->abrt_row, "active",
+                   G_SETTINGS_BIND_DEFAULT);
+
+#if defined(HAVE_WHOOPSIE) || defined(HAVE_INSIGHTS)
+  self->cancellable = g_cancellable_new ();
+#endif
+
+#ifdef HAVE_WHOOPSIE
+  /* check for whoopsie */
+  whoopsie_preferences_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
+                                          G_DBUS_PROXY_FLAGS_NONE,
+                                          "com.ubuntu.WhoopsiePreferences",
+                                          "/com/ubuntu/WhoopsiePreferences",
+                                          self->cancellable,
+                                          on_new_whoopsie_proxy_cb,
+                                          self);
+
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->abrt_row), FALSE);
+  gtk_widget_set_sensitive (GTK_WIDGET (self->whoopsie_combo_row), FALSE);
+#endif
+
+#ifdef HAVE_INSIGHTS
+  self->insights_cancellable = g_cancellable_new ();
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_group), TRUE);
+
+  self->permission = polkit_permission_new_sync (INSIGHTS_SYSTEM_OPT_OUT_PERMISSION,
+                                                 NULL, NULL, &error);
+  if (self->permission == NULL)
+    {
+      g_critical ("Cannot create '%s' permission: %s",
+                  INSIGHTS_SYSTEM_OPT_OUT_PERMISSION, error->message);
+      g_clear_error (&error);
+    }
+  else
+    {
+      gtk_widget_set_visible (GTK_WIDGET (self->permission_infobar), TRUE);
+      cc_permission_infobar_set_permission (self->permission_infobar,
+                                            self->permission);
+    }
+
+  /* Unlocking via the infobar changes "allowed"; keep the switch's
+   * sensitivity (and tooltip) in sync. */
+  if (self->permission != NULL)
+    g_signal_connect_object (self->permission, "notify::allowed",
+                             G_CALLBACK (sync_insights_switches),
+                             self, G_CONNECT_SWAPPED);
+
+  g_signal_connect (self, "map", G_CALLBACK (sync_insights_switches), NULL);
+  g_signal_connect (self, "unmap", G_CALLBACK (on_unmap_cb), NULL);
+#endif
+
+  os_name = g_get_os_info (G_OS_INFO_KEY_NAME);
+  if (!os_name)
+    os_name = g_strdup ("GNOME");
+  whoopsie_url = g_get_os_info (G_OS_INFO_KEY_PRIVACY_POLICY_URL);
+  if (whoopsie_url)
+    gtk_link_button_set_uri (self->diagnostics_learn_more_link_button, whoopsie_url);
+
+  /* translators: The '%s' is the distributor's name, such as 'Fedora' */
+  whoopsie_msg = g_strdup_printf (_ ("Help us address technical problems by sharing error reports with the "
+                                     "%s team. Reports are sent anonymously and are scrubbed of personal data."),
+                                  os_name);
+  adw_preferences_group_set_description (self->diagnostics_group, whoopsie_msg);
+
+  /* translators: The '%s' is the distributor's name, such as 'Ubuntu' */
+  insights_msg = g_strdup_printf (_ ("Help us improve %s features and compatibility by sharing system reports with"
+                                     " Canonical. Reports are sent anonymously and do not contain any personal data."),
+                                  os_name);
+
+  adw_preferences_group_set_description (self->insights_group, insights_msg);
+
+  provider = gtk_css_provider_new ();
+  gtk_css_provider_load_from_resource (provider, "/org/gnome/control-center/privacy/telemetry/telemetry.css");
+  gtk_style_context_add_provider_for_display (gdk_display_get_default (),
+                                              GTK_STYLE_PROVIDER (provider),
+                                              GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+}
diff --git a/panels/privacy/diagnostics/cc-diagnostics-page.h b/panels/privacy/telemetry/cc-telemetry-page.h
similarity index 73%
copy from panels/privacy/diagnostics/cc-diagnostics-page.h
copy to panels/privacy/telemetry/cc-telemetry-page.h
index 22dadff..acb5f1a 100644
--- a/panels/privacy/diagnostics/cc-diagnostics-page.h
+++ b/panels/privacy/telemetry/cc-telemetry-page.h
@@ -1,5 +1,6 @@
 /* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
  *
+ * Copyright (C) 2026 Canonical Ltd.
  * Copyright (C) 2018 Red Hat, Inc
  *
  * This program is free software; you can redistribute it and/or modify
@@ -15,7 +16,9 @@
  * You should have received a copy of the GNU General Public License
  * along with this program; if not, see <http://www.gnu.org/licenses/>.
  *
- * Author: Matthias Clasen <mclasen@redhat.com>
+ * Author(s):
+ *   Kat Kuo <kat.kuo@canonical.com>
+ *   Matthias Clasen <mclasen@redhat.com>
  */
 
 #pragma once
@@ -24,6 +27,7 @@
 
 G_BEGIN_DECLS
 
-#define CC_TYPE_DIAGNOSTICS_PAGE (cc_diagnostics_page_get_type ())
-G_DECLARE_FINAL_TYPE (CcDiagnosticsPage, cc_diagnostics_page, CC, DIAGNOSTICS_PAGE, AdwNavigationPage);
+#define CC_TYPE_TELEMETRY_PAGE (cc_telemetry_page_get_type ())
+G_DECLARE_FINAL_TYPE (CcTelemetryPage, cc_telemetry_page, CC, TELEMETRY_PAGE, AdwNavigationPage)
+
 G_END_DECLS
diff --git a/panels/privacy/telemetry/gnome-control-center-insights-helper.c b/panels/privacy/telemetry/gnome-control-center-insights-helper.c
new file mode 100644
index 0000000..83900da
--- /dev/null
+++ b/panels/privacy/telemetry/gnome-control-center-insights-helper.c
@@ -0,0 +1,62 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2026 Canonical Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Privileged helper that sets the Ubuntu Insights system-wide opt-out state.
+ * Intended to be launched via pkexec, which enforces the
+ * org.gnome.controlcenter.insights-system-opt-out polkit action. */
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <insights/insights.h>
+
+int
+main (int   argc,
+      char *argv[])
+{
+  char *error_msg = NULL;
+  bool  state;
+
+  if (argc != 2)
+    {
+      fprintf (stderr, "Usage: %s true|false\n", argv[0]);
+      return EXIT_FAILURE;
+    }
+
+  if (strcmp (argv[1], "true") == 0)
+    state = true;
+  else if (strcmp (argv[1], "false") == 0)
+    state = false;
+  else
+    {
+      fprintf (stderr, "%s: invalid state '%s', expected true or false\n", argv[0], argv[1]);
+      return EXIT_FAILURE;
+    }
+
+  error_msg = insights_set_system_opt_out_state (NULL, state);
+  if (error_msg != NULL)
+    {
+      fprintf (stderr, "Failed to set system opt-out state: %s\n", error_msg);
+      free (error_msg);
+      return EXIT_FAILURE;
+    }
+
+  return EXIT_SUCCESS;
+}
diff --git a/panels/privacy/telemetry/meson.build b/panels/privacy/telemetry/meson.build
new file mode 100644
index 0000000..ee416bf
--- /dev/null
+++ b/panels/privacy/telemetry/meson.build
@@ -0,0 +1,42 @@
+ui_sources = files(
+  'cc-telemetry-page.blp',
+)
+
+if enable_insights
+  ui_sources += files('cc-telemetry-insights-example.blp')
+endif
+
+ui_compiled = []
+foreach ui_source : ui_sources
+  ui_compiled += fs.name(fs.replace_suffix(ui_source, '.ui'))
+endforeach
+
+blueprints += custom_target(
+  input: ui_sources,
+  output: ui_compiled,
+  command: [blueprint_compiler, 'batch-compile', '@OUTDIR@', '@CURRENT_SOURCE_DIR@', '@INPUT@']
+)
+
+if enable_insights
+  executable(
+    'gnome-control-center-insights-helper',
+    sources: ['gnome-control-center-insights-helper.c'],
+    dependencies: insights_dep,
+    install: true,
+    install_dir: control_center_libexecdir,
+  )
+
+  i18n.merge_file(
+    input: configure_file(
+      input: 'org.gnome.controlcenter.insights-system-opt-out.policy.in.in',
+      output: '@BASENAME@',
+      configuration: {
+        'libexecdir': control_center_libexecdir,
+      }
+    ),
+    output: '@BASENAME@',
+    po_dir: po_dir,
+    install: true,
+    install_dir: join_paths(control_center_datadir, 'polkit-1', 'actions')
+  )
+endif
diff --git a/panels/keyboard/org.gnome.controlcenter.keyboard.policy.in.in b/panels/privacy/telemetry/org.gnome.controlcenter.insights-system-opt-out.policy.in.in
similarity index 60%
copy from panels/keyboard/org.gnome.controlcenter.keyboard.policy.in.in
copy to panels/privacy/telemetry/org.gnome.controlcenter.insights-system-opt-out.policy.in.in
index 7f550f3..511d8c2 100644
--- a/panels/keyboard/org.gnome.controlcenter.keyboard.policy.in.in
+++ b/panels/privacy/telemetry/org.gnome.controlcenter.insights-system-opt-out.policy.in.in
@@ -2,21 +2,20 @@
 <!DOCTYPE policyconfig PUBLIC
  "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
  "http://www.freedesktop.org/standards/PolicyKit/1.0/policyconfig.dtd">
-
 <policyconfig>
+
   <vendor>The GNOME Project</vendor>
   <vendor_url>http://www.gnome.org/</vendor_url>
 
-  <action id="org.gnome.controlcenter.set-keyboard">
-    <description>Change system keyboard settings</description>
-    <message>To change global keyboard settings, you need to authenticate.</message>
+  <action id="org.gnome.controlcenter.insights-system-opt-out">
+    <description>Change system-wide telemetry reporting</description>
+    <message>Authentication is required to change system reporting for all users</message>
     <defaults>
       <allow_any>no</allow_any>
       <allow_inactive>no</allow_inactive>
       <allow_active>auth_admin_keep</allow_active>
     </defaults>
-    <annotate key="org.freedesktop.policykit.exec.path">@libexecdir@/gnome-control-center-set-debian-keyboard</annotate>
-    <annotate key="org.freedesktop.policykit.imply">org.freedesktop.locale1.set-keyboard</annotate>
+    <annotate key="org.freedesktop.policykit.exec.path">@libexecdir@/gnome-control-center-insights-helper</annotate>
   </action>
 
 </policyconfig>
diff --git a/panels/privacy/telemetry/telemetry.css b/panels/privacy/telemetry/telemetry.css
new file mode 100644
index 0000000..281d579
--- /dev/null
+++ b/panels/privacy/telemetry/telemetry.css
@@ -0,0 +1,8 @@
+/* Emulate spacing of a label */
+.link-label {
+  padding: 0;
+  box-shadow: none;
+  border-radius: 0px;
+  min-height: 0px;
+  min-width: 0px;
+}
diff --git a/po/POTFILES.in b/po/POTFILES.in
index 89927a7..30191cc 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -173,6 +173,12 @@ panels/privacy/firmware-security/cc-firmware-security-utils.c
 panels/privacy/gnome-privacy-panel.desktop.in
 panels/privacy/location/cc-location-page.blp
 panels/privacy/screen/cc-screen-page.blp
+panels/privacy/telemetry/cc-telemetry-insights-example.blp
+panels/privacy/telemetry/cc-telemetry-insights-example.c
+panels/privacy/telemetry/cc-telemetry-page.blp
+panels/privacy/telemetry/cc-telemetry-page.c
+panels/privacy/telemetry/gnome-control-center-insights-helper.c
+panels/privacy/telemetry/org.gnome.controlcenter.insights-system-opt-out.policy.in.in
 panels/privacy/usage/cc-usage-page.blp
 panels/search/cc-search-locations-page.blp
 panels/search/cc-search-locations-page.c
